image-copy/cmd/app/main.go

```diff
@@ -0,0 +1,172 @@
+	Port     int    `envconfig:"PORT" default:"8080" required:"true"`
+	DstRepo  string `envconfig:"DST_REPO" required:"true"` // Almost fully qualified at this point, just needs the final component.
+}
+
+var location, sa string
+var srcRepo name.Repository
+
+func init() {
+	var err error
+	location, err = metadata.Zone()
+	if err != nil {
+		log.Panicf("getting location: %v", err)
+	}
+	sa, err = metadata.Email("default")
+	if err != nil {
+		log.Panicf("getting SA: %v", err)
+	}
+}
+
+func main() {
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		log.Fatalf("failed to process env var: %s", err)
+	}
+	log.Printf("env: %+v", env)
+	log.Printf("location: %s", location)
+	log.Printf("sa: %s", sa)
+
+	c, err := cloudevents.NewClientHTTP(cloudevents.WithPort(env.Port),
+		// We need to infuse the request onto context, so we can
+		// authenticate requests.
+		cehttp.WithRequestDataAtContextMiddleware())
+	if err != nil {
+		log.Fatalf("failed to create client, %v", err)
+	}
+	ctx := context.Background()
+
+	// Construct a verifier that ensures tokens are issued by the Chainguard
+	// issuer we expect and are intended for a customer webhook.
+	provider, err := oidc.NewProvider(ctx, env.Issuer)
+	if err != nil {
+		log.Fatalf("failed to create provider: %v", err)
+	}
+	verifier := provider.Verifier(&oidc.Config{
+		ClientID: "customer",
+	})
+
+	receiver := func(ctx context.Context, event cloudevents.Event) error {
+		// We expect Chainguard webhooks to pass an Authorization header.
+		auth := strings.TrimPrefix(cehttp.RequestDataFromContext(ctx).Header.Get("Authorization"), "Bearer ")
+		if auth == "" {
+			return cloudevents.NewHTTPResult(http.StatusUnauthorized, "Unauthorized")
+		}
+
+		// Verify that the token is well-formed, and in fact intended for us!
+		if tok, err := verifier.Verify(ctx, auth); err != nil {
+			return cloudevents.NewHTTPResult(http.StatusForbidden, "unable to verify token: %w", err)
+		} else if !strings.HasPrefix(tok.Subject, "webhook:") {
+			return cloudevents.NewHTTPResult(http.StatusForbidden, "subject should be from the Chainguard webhook component, got: %s", tok.Subject)
+		} else if group := strings.TrimPrefix(tok.Subject, "webhook:"); group != env.Group {
+			return cloudevents.NewHTTPResult(http.StatusForbidden, "this token is intended for %s, wanted one for %s", group, env.Group)
+		}
+
+		// We are handling a specific event type, so filter the rest.
+		if event.Type() != PushEventType {
+			return nil
+		}
+
+		data := Occurrence{}
+		if err := event.DataAs(&data); err != nil {
+			return cloudevents.NewHTTPResult(http.StatusInternalServerError, "unable to unmarshal data: %w", err)
+		}
+
+		log.Printf("got event: %+v", data)
+
+		src := "cgr.dev/" + data.Body.Repository
+		dst := env.DstRepo + "/" + filepath.Base(data.Body.Repository)
+		log.Printf("Copying %s to %s...", src, dst)
+		if err := crane.Copy(src, dst,
+			crane.WithAuthFromKeychain(authn.NewMultiKeychain(
```

Comment:
NB: I know that technically this will catch references because those will get their own push events, but I'd love it if we were using `cosign copy` here.

Reply:
I agree, I can do this in a followup.

If there was a `cosign.Copy` that didn't bring in a world of dependencies I probably would have used that from the outset. 😒